src/Symfony/Component/ObjectMapper/ObjectMapper.php

```diff
@@ -0,0 +1,326 @@
+     */
+    public function __construct(
+        private readonly ObjectMapperMetadataFactoryInterface $metadataFactory = new ReflectionObjectMapperMetadataFactory(),
+        private readonly ?PropertyAccessorInterface $propertyAccessor = null,
+        private readonly ?ContainerInterface $transformCallableLocator = null,
+        private readonly ?ContainerInterface $conditionCallableLocator = null,
+    ) {
+    }
+
+    public function map(object $source, object|string|null $target = null): object
+    {
+        static $objectMap = null;
+        $objectMapInitialized = false;
+
+        if (null === $objectMap) {
+            $objectMap = new \SplObjectStorage();
+            $objectMapInitialized = true;
+        }
+
+        $metadata = $this->metadataFactory->create($source);
+        $map = $this->getMapTarget($metadata, null, $source);
+        $target ??= $map?->target;
+        $mappingToObject = \is_object($target);
+
+        if (!$target) {
+            throw new MappingException('Mapping target not found.');
+        }
+
+        if (\is_string($target) && !class_exists($target)) {
+            throw new MappingException(\sprintf('Mapping target "%s" not found.', $target));
+        }
+
+        try {
+            $targetRefl = new \ReflectionClass($target);
+        } catch (\ReflectionException $e) {
+            throw new MappingException($e->getMessage(), $e->getCode(), $e);
+        }
+
+        $mappedTarget = $mappingToObject ? $target : $targetRefl->newInstanceWithoutConstructor();
+        if ($map && $map->transform) {
+            $mappedTarget = $this->applyTransforms($map, $mappedTarget, $mappedTarget);
+
+            if (!\is_object($mappedTarget)) {
+                throw new MappingTransformException('Cannot map to a non-object.');
+            }
+        }
+
+        if (!is_a($mappedTarget, $targetRefl->getName(), false)) {
+            throw new MappingException(\sprintf('Expected the mapped object to be an instance of "%s".', $mappingToObject ? $mappedTarget::class : $mappedTarget));
+        }
+
+        $objectMap[$source] = $mappedTarget;
+        $ctorArguments = [];
+        $constructor = $targetRefl->getConstructor();
+        foreach ($constructor?->getParameters() ?? [] as $parameter) {
+            $parameterName = $parameter->getName();
+            if (!$targetRefl->hasProperty($parameterName)) {
+                continue;
+            }
+
+            $property = $targetRefl->getProperty($parameterName);
+
+            // The mapped class was probably instantiated in a transform we can't write a readonly property
+            if ($property->isReadOnly() && ($property->isInitialized($mappedTarget) && $property->getValue($mappedTarget))) {
+                continue;
+            }
+
+            $ctorArguments[$parameterName] = $parameter->isDefaultValueAvailable() ? $parameter->getDefaultValue() : null;
```

Comment:
ideally, we wouldn't call getDefaultValue
can't using named arguments prevent doing it?